full/runner.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2017 Palantir Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package full
+
+import (
+	"os"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/pkg/errors"
+
+	"github.com/palantir/bouncer/bouncer"
+)
+
+// Runner holds data for a particular full run
+// Note that in the full case, asgs will always be of length 1
+type Runner struct {
+	bouncer.BaseRunner
+}
+
+// NewRunner instantiates a new full runner
+func NewRunner(opts *bouncer.RunnerOpts) (*Runner, error) {
+	br, err := bouncer.NewBaseRunner(opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting base runner")
+	}
+
+	r := Runner{
+		*br,
+	}
+	return &r, nil
+}
+
+// MustValidatePrereqs checks that the batch runner is safe to proceed
+func (r *Runner) MustValidatePrereqs() {
+	asgSet, err := r.NewASGSet()
+	if err != nil {
+		log.Fatal(errors.Wrap(err, "error building ASGSet"))
+	}
+
+	divergedASGs := asgSet.GetDivergedASGs()
+	if len(divergedASGs) != 0 {
+		for _, badASG := range divergedASGs {
+			log.WithFields(log.Fields{
+				"ASG":                     *badASG.ASG.AutoScalingGroupName,
+				"desired_capacity actual": *badASG.ASG.DesiredCapacity,
+				"desired_capacity given":  badASG.DesiredASG.DesiredCapacity,
+			}).Error("ASG desired capacity doesn't match expected starting value")
```

Comment:
So, this check might not be relevant in your case.

In the serial & canary cases, the idea was, if bouncer failed, you the user would kick-off the TF run again and let TF set the descap back to what its steady-state value is, and then bouncer wouldn't proceed until that had happened. It was really just a safeguard.

In your case, I'm not sure if that's what you'd do or not? If your bouncer run fails and you go to re-run, what's expected, that you'd hit TF again? If so, it would rehydrate the drained ASGs by setting their descaps back to the steady-state. If this is fine, then this check is fine.

If, in that situation, you'd instead expect to just run bouncer again without running TF again, then I think you'll want to remove this prereq check.

Reply:
I think the same safeguard makes sense.  If the run does fail, we'd want the same guarantee that things would be set back to the normal run state, as this new mode is essentially the same as serial, with a different order of operations.

Reply:
Cool.